plutus-core/plutus-ir/test/TypecheckSpec.hs

```diff
@@ -0,0 +1,141 @@
+  , testProperty "strictifyBindings (Builtin Variant 2) pass" $
+  withMaxSuccess (factor*3000) (prop_typecheck_strictifyBindings DefaultFunSemanticsVariant2)
+  , testProperty "evaluateBuiltins (Builtin Variant1) pass" $
+  withMaxSuccess (factor*3000) (prop_typecheck_evaluateBuiltins DefaultFunSemanticsVariant1)
+  , testProperty "evaluateBuiltins (Builtin Variant2) pass" $
+  withMaxSuccess (factor*3000) (prop_typecheck_evaluateBuiltins DefaultFunSemanticsVariant2)
+  -- FIXME
+  , expectFail $ testProperty "inline (Builtin Variant1) pass" $
+  withMaxSuccess (factor*3000) (prop_typecheck_inline DefaultFunSemanticsVariant1)
+  -- FIXME
+  , expectFail $ testProperty "inline (Builtin Variant2) pass" $
+  withMaxSuccess (factor*3000) (prop_typecheck_inline DefaultFunSemanticsVariant2)
+  ]
+
+-- Convert Either Error () to Either String () to match with the Testable (Either String ())
+-- instance.
+convertToEitherString :: Either (Error DefaultUni DefaultFun ()) ()
+  -> Either String ()
+convertToEitherString = \case
+  Left err -> Left $ show err
+  Right () -> Right ()
+
+-- | Typechecking a term after one of the pure passes.
+tyCheckPure ::
+  -- | The pure simplification pass.
+  (Term TyName Name DefaultUni DefaultFun () -> Term TyName Name DefaultUni DefaultFun ())
+  -> Term TyName Name DefaultUni DefaultFun ()
+  -> Either (Error DefaultUni DefaultFun ()) ()
+tyCheckPure pass tm = do
+  config <- getDefTypeCheckConfig ()
+  inferredType <- runQuoteT $ inferType config tm
+  runQuoteT $ checkType config () (pass tm) inferredType
+
+prop_typecheck_pure ::
+  (Term TyName Name DefaultUni DefaultFun ()
+  -> Term TyName Name DefaultUni DefaultFun ())
+  -> Property
+prop_typecheck_pure pass =
+  -- generate type and term in debug mode for easier debugging
+  forAllDoc "ty,tm" genTypeAndTermDebug_ (const []) $ \ (_ty, tm) -> do
+  convertToEitherString $ tyCheckPure pass tm
+
+-- | Typechecking a term after a `PlutusIR.Compiler.simplifyTerm` pass.
+prop_typecheck :: Property
+prop_typecheck =
+  -- generate type and term in debug mode for easier debugging
+  forAllDoc "ty,tm" genTypeAndTermDebug_ (const []) $ \ (_ty, tm) ->
+    convertToEitherString $ do
+    config <- getDefTypeCheckConfig ()
+    plcConfig <- PLC.getDefTypeCheckConfig ()
+    inferredType <- runQuoteT $ inferType config tm
+    case runReaderT (runQuoteT (simplifyTerm (original tm))) (toDefaultCompilationCtx plcConfig) of
+      Left err -> Left $ void err
+      Right simplified ->
+        runQuoteT $ checkType config () (void simplified) inferredType
+
+-- | Typechecking a term after a `PlutusIR.Transform.StrictifyBindings` pass.
+prop_typecheck_strictifyBindings :: BuiltinSemanticsVariant DefaultFun -> Property
+prop_typecheck_strictifyBindings biVariant =
+  -- generate type and term in debug mode for easier debugging
+  forAllDoc "ty,tm" genTypeAndTermDebug_ (const []) $ \ (_ty, tm) ->
```

Comment:
It seems to me that these all fit into `prop_typecheck_pure`, you just have to capture the other stuff in the pass function, e.g.
```
prop_typecheck_strictifyBindings biVariant = prop_typecheck_pure (strictifyBindings (BuiltinsInfo biVariant)
```

Reply:
Oops yeah only `inline` is not pure actually.